test/functional/test_framework/test_node.py

```diff
@@ -307,21 +307,31 @@ def wait_until_stopped(self, timeout=BITCOIND_PROC_WAIT_TIMEOUT):
         wait_until(self.is_node_stopped, timeout=timeout)
 
     @contextlib.contextmanager
-    def assert_debug_log(self, expected_msgs):
-        debug_log = os.path.join(self.datadir, self.chain, 'debug.log')
-        with open(debug_log, encoding='utf-8') as dl:
-            dl.seek(0, 2)
-            prev_size = dl.tell()
-        try:
-            yield
-        finally:
+    def assert_debug_log(self, expected_msgs, timeout=2):
+        time_end = time.time() + timeout
+        print_log = ""
+        while True:
+            debug_log = os.path.join(self.datadir, self.chain, 'debug.log')
+            found = True
             with open(debug_log, encoding='utf-8') as dl:
-                dl.seek(prev_size)
-                log = dl.read()
-            print_log = " - " + "\n - ".join(log.splitlines())
-            for expected_msg in expected_msgs:
-                if re.search(re.escape(expected_msg), log, flags=re.MULTILINE) is None:
-                    self._raise_assertion_error('Expected message "{}" does not partially match log:\n\n{}\n\n'.format(expected_msg, print_log))
+                dl.seek(0, 2)
+                prev_size = dl.tell()
```

Comment:
It looks like you are resetting prev_size in a loop, thus introducing another race where we just miss the debug log lines between .read() and .tell(), no?